tree: block interval::string for Stable cases when IntervalStyle_enabled

otan opened the pull request:
See individual commits for details.

Resolves #67887

otan committed:
builtins: implement parse_interval and to_char_with_style

Release note (sql change): Implement the parse_interval and
to_char_with_style builtins, which converts strings from/to intervals
with immutable volatility.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67970)
<!-- Reviewable:end -->

otan committed:
tree: block interval::string (and vice versa) when IntervalStyle_enabled

Release note (sql change): Casting from interval to string or vice versa
is now blocked for computed columns, partial indexes and partitions when
the intervalstyle_enabled session setting is enabled. Instead, using
`to_char_with_style(interval, style)` or
`parse_interval(interval, intervalstyle)` is available as a substitute.
This is enforced in v22.1, but is opt-in for v21.2. It is recommended to
set the CLUSTER SETTING `sql.defaults.intervalstyle_enabled` to true to
avoid surprises when upgrading to v22.1.

rafiss reviewed (approved):
this lgtm!

this idea is super late to the game, but i just thought of one more idea: if we made a rule that `SET intervalstyle` is not allowed during a session but you can only configure intervalstyle during session initialization, would that let us keep the cast as immutable all the time?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67970#-:-Mf_yWEZ4MBgNKOeo_Bk:byx3l4e)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

otan commented:
i think it's still risky for e.g. computed columns, where the stability is important in the context of the session style. even if it's a session initialization thing, the real problem lies in the partial indexes and the like.

i'm happy with the rules we have setup for now.

TFTR!

bors r=rafiss
